src/lib/PerceptronPredictor.cc

```diff
@@ -0,0 +1,159 @@
+
+  // Determine direction prediction from perceptron by taking its dot product
+  // with the global history
+  // Starting with the bias weight
+  int64_t Pout = perceptron[globalHistoryLength_];
+  for (int i = 0; i < globalHistoryLength_; i++) {
+    // Get branch direction for ith entry in the globalHistory_
+    bool historyTaken =
+        ((globalHistory_ & (1 << ((globalHistoryLength_ - 1) - i))) != 0);
+    Pout += historyTaken ? perceptron[i] : (0 - perceptron[i]);
+  }
+  bool direction = (Pout >= 0);
+
+  // Retrieve target prediction
+  uint64_t target =
+      (knownOffset != 0) ? address + knownOffset : btb_[hashedIndex].second;
+
+  BranchPrediction prediction = {direction, target};
+
+  // Amend prediction based on branch type
+  if (type == BranchType::Unconditional) {
+    prediction.taken = true;
+  } else if (type == BranchType::Return) {
+    prediction.taken = true;
+    // Return branches can use the RAS if an entry is available
+    if (ras_.size() > 0) {
+      prediction.target = ras_.back();
+      // Record top of RAS used for target prediction
+      rasHistory_[address] = ras_.back();
+      ras_.pop_back();
+    }
+  } else if (type == BranchType::SubroutineCall) {
+    prediction.taken = true;
+    // Subroutine call branches must push their associated return address to RAS
+    if (ras_.size() >= rasSize_) {
+      ras_.pop_front();
+    }
+    ras_.push_back(address + 4);
+    // Record that this address is a branch-and-link instruction
+    rasHistory_[address] = 0;
+  } else if (type == BranchType::Conditional) {
+    if (!prediction.taken) prediction.target = address + 4;
+  }
+  return prediction;
+}
+
+void PerceptronPredictor::update(uint64_t address, bool taken,
+                                 uint64_t targetAddress, BranchType type) {
+  // Work out hash index
+  uint64_t prevGlobalHistory = btbHistory_[address];
+  uint64_t hashedIndex =
+      ((address >> 2) ^ prevGlobalHistory) & ((1 << btbBits_) - 1);
+
+  std::vector<int8_t> perceptron = btb_[hashedIndex].first;
+
+  // Work out the most recent prediction
+  int64_t Pout = perceptron[globalHistoryLength_];
+  for (int i = 0; i < globalHistoryLength_; i++) {
+    bool historyTaken =
+        ((prevGlobalHistory & (1 << ((globalHistoryLength_ - 1) - i))) != 0);
+    Pout += historyTaken ? perceptron[i] : (0 - perceptron[i]);
+  }
+  bool directionPrediction = (Pout >= 0);
+
+  // Determine the magnitude of the dot product for training
+  uint64_t magnitude = abs(Pout);
+
+  // Update the perceptron if the prediction was wrong, or the dot product's magnitude
+  // was not greater than the training threshold
+  if ((directionPrediction != taken) || (magnitude < trainingThreshold_)) {
+    int8_t t = (taken) ? 1 : -1;
+
+    for (int i = 0; i < globalHistoryLength_; i++) {
+      int8_t xi =
+          ((prevGlobalHistory & (1 << ((globalHistoryLength_ - 1) - i))) == 0) ? -1 : 1;
+      int8_t product_xi_t = xi * t;
+      // Make sure no overflow
```

Comment:
Perhaps change the comment to "Make sure no overflow, and cap out perceptron weight to +- 127". Took me a second to think through the logic of what was happening if we do say reach 127 with a +1 movement, and this would clear it up.